eyeball-im-util/src/vector/limit.rs

```diff
@@ -0,0 +1,361 @@
+        match diff {
+            VectorDiff::Append { values } => self.buffered_vector.append(values.clone()),
+            VectorDiff::Clear => self.buffered_vector.clear(),
+            VectorDiff::PushFront { value } => self.buffered_vector.push_front(value.clone()),
+            VectorDiff::PushBack { value } => self.buffered_vector.push_back(value.clone()),
+            VectorDiff::PopFront => {
+                self.buffered_vector.pop_front();
+            }
+            VectorDiff::PopBack => {
+                self.buffered_vector.pop_back();
+            }
+            VectorDiff::Insert { index, value } => {
+                self.buffered_vector.insert(*index, value.clone());
+            }
+            VectorDiff::Set { index, value } => {
+                self.buffered_vector.set(*index, value.clone());
+            }
+            VectorDiff::Remove { index } => {
+                self.buffered_vector.remove(*index);
+            }
+            VectorDiff::Truncate { length } => self.buffered_vector.truncate(*length),
+            VectorDiff::Reset { values } => {
+                *self.buffered_vector = values.clone();
+            }
+        }
+    }
+
+    /// Update the limit if necessary.
+    ///
+    /// * If the buffered vector is empty, it returns `None`.
+    /// * If the limit increases, a `VectorDiff::Append` is produced if any
+    ///   items exist.
+    /// * If the limit decreases below the length of the vector, a
+    ///   `VectorDiff::Truncate` is produced.
+    ///
+    /// It's OK to have a `new_limit` larger than the length of the `Vector`.
+    /// The `new_limit` won't be capped.
+    fn update_limit(&mut self, next_limit: usize) -> Option<S::Item> {
+        if self.buffered_vector.is_empty() {
+            // Let's update the limit.
+            *self.limit = next_limit;
+
+            // But let's ignore any update.
+            return None;
+        }
+
+        match (*self.limit).cmp(&next_limit) {
+            // old < new
+            Ordering::Less => {
+                // Let's add the missing items.
+                let mut missing_items = vector![];
+
+                for nth in *self.limit..min(self.buffered_vector.len(), next_limit) {
+                    // SAFETY: It's OK to `unwrap` here as we are sure we are iterating over defined
+                    // items.
+                    let item = self.buffered_vector.get(nth).unwrap();
+                    missing_items.push_back(item.clone());
+                }
+
+                // Let's update the limit.
+                *self.limit = next_limit;
+
+                let missing_items =
+                    S::Item::from_item(VectorDiff::Append { values: missing_items });
+
+                Some(missing_items)
+            }
+
+            // old > new
+            Ordering::Greater => {
+                // Let's remove the extra items.
+                let items_removal = S::Item::from_item(VectorDiff::Truncate { length: next_limit });
```

Comment:
We can return `None` instead if `buffered_vector.len() <= new_limit`.

Reply:
Indeed! Good catch. I've updated the tests to test this case in particular.